atc/worker/pool.go

```diff
@@ -174,40 +286,86 @@ func (pool *pool) SelectWorker(
 	containerSpec ContainerSpec,
 	workerSpec WorkerSpec,
 	strategy ContainerPlacementStrategy,
-) (Client, error) {
+	callbacks PoolCallbacks,
+) (Client, time.Duration, error) {
 	logger := lagerctx.FromContext(ctx)
 
-	workersWithContainer, err := pool.provider.FindWorkersForContainerByOwner(
-		logger.Session("find-worker"),
-		owner,
-	)
-	if err != nil {
-		return nil, err
+	started := time.Now()
+	labels := metric.StepsWaitingLabels{
+		Platform:   workerSpec.Platform,
+		TeamId:     strconv.Itoa(workerSpec.TeamID),
+		Type:       string(containerSpec.Type),
+		WorkerTags: strings.Join(workerSpec.Tags, "_"),
 	}
 
-	compatibleWorkers, err := pool.allSatisfying(logger, workerSpec)
-	if err != nil {
-		return nil, err
-	}
+	var worker Client
+	var pollingTicker *time.Ticker
+	for {
+		var err error
+		worker, err = pool.findWorker(ctx, owner, containerSpec, workerSpec, strategy)
 
-	var worker Worker
-dance:
-	for _, w := range workersWithContainer {
-		for _, c := range compatibleWorkers {
-			if w.Name() == c.Name() {
-				worker = c
-				break dance
+		if err != nil {
+			return nil, 0, err
+		}
+
+		if worker != nil {
+			break
+		}
+
+		if pollingTicker == nil {
+			pollingTicker = time.NewTicker(WorkerPollingInterval)
+			defer pollingTicker.Stop()
+
+			logger.Debug("waiting-for-available-worker")
+
+			_, ok := metric.Metrics.StepsWaiting[labels]
+			if !ok {
+				metric.Metrics.StepsWaiting[labels] = &metric.Gauge{}
+			}
+
+			metric.Metrics.StepsWaiting[labels].Inc()
+			defer metric.Metrics.StepsWaiting[labels].Dec()
+
+			if callbacks != nil {
+				callbacks.WaitingForWorker(logger)
 			}
 		}
-	}
 
-	if worker == nil {
-		worker, err = strategy.Choose(logger, compatibleWorkers, containerSpec)
-		if err != nil {
-			return nil, err
+		select {
+		case <-ctx.Done():
+			logger.Info("aborted-waiting-for-worker")
+			return nil, 0, ctx.Err()
+		case <-pollingTicker.C:
+		case <-pool.waker:
+			break
```

Comment:
Can we get rid of this `break`? At first glance, I was wondering why it'd behave differently from `pollingTicker.C`, but then realized it doesn't since the `break` is a no-op (it just breaks out of the `select`, which it would do anyway)